### PR TITLE
[@types/which] Update the options of `which` for version 3.0.0.

### PR DESCRIPTION
**Pull request by vthemelis**
Add `nothrow` option to both the sync and async APIs and add the missing `delimiter` option.

These have been available on the package since version 3.0.0: https://github.com/npm/node-which/commit/8b0187ceab57b0814ad6a77a5706319ffa5bf103

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test which`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/npm/node-which/blob/main/lib/index.js
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

**Comment by vthemelis** [comment]
The issue with this is that something like:
```typescript
which('blah', { not_a_real_option: true })
```

typechecks fine (when it should probably not). Not totally sure how to avoid this just now. Would you recommend that I add more overloads instead?

**Review by EduardoAC — APPROVED** [review]
LGTM

**Comment by vthemelis** [comment]
@vvakame, could you have a look? I think that there is parity in terms of what does and doesn't compiler now with the previous version (excluding the intentionally introduced new behaviour).

**Comment by vthemelis** [comment]
Cc @EduardoAC @peterblazejewicz

**Comment by vthemelis** [comment]
CC @cspotcode

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/which/index.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 78b9e76dfa0e94b2f6c48741652d5b1db753a226
</p>

**Comment by vthemelis** [comment]
@EduardoAC can I have another review? I pushed some more changes to completely address typing for the latest version of the package.

File: types/which/index.d.ts
Comment (EduardoAC) [inline types/which/index.d.ts:13]: Can you expand on the naming for AddNull and AdddArray? I see that the new type a bit obscure to understand what it represents.

When i see the code feels that you are more append or transforming the original types

Comment (vthemelis) [reply types/which/index.d.ts:13]: Sure, happy to do that. What name would work best? Maybe NullableIfNoThrow and ArrayIfAll ?

Comment (EduardoAC) [reply types/which/index.d.ts:13]: > Sure, happy to do that. What name would work best? Maybe NullableIfNoThrow and ArrayIfAll ?

I was thinking something quite similar. Something in lines`AppendNullIfNothrow` and `TransformToArrayIfAll`. What are your thoughts? If you prefer yours, i will accept them as well

Comment (vthemelis) [reply types/which/index.d.ts:13]: Sounds good, will amend later.

Comment (vthemelis) [reply types/which/index.d.ts]: Done!

**Review by EduardoAC — APPROVED** [review]
LGTM

File: types/which/index.d.ts
Comment (EduardoAC) [reply types/which/index.d.ts]: Well done, i appreciated the improvements

**Comment by vthemelis** [comment]
Thanks for the review @EduardoAC ! Do I need to get another review before merging?

**Comment by EduardoAC** [comment]
> Thanks for the review @EduardoAC ! Do I need to get another review before merging?

Usually, the merge is done by the repository owner, but if you can merge, go ahead. LGTM

**Review by alvinsga — APPROVED** [review]

**Review by sheetalkamat — APPROVED** [review]

**Comment by vthemelis** [comment]
Ready to merge

**Pull request merged**
